drivers: eeprom: implement driver st25dv

nono313 opened the pull request:
Based on #35842 I have created this new PR to get an eeprom driver for the st25dv into Zephyr.

Tasks done in this PR:

- Copied code from original PR
- Implemented (some) specific features of the st25dv, including user area configuration, UUID, IC rev / ref read, RF protection configuration
- Updated the code to match current Zephyr version
- Implemented a shield X-NUCLEO-NFC04A1 that has 3 LEDs and a st25dv chip on it
- Wrote a sample project that uses both the EEPROM and the LEDs on the X-NUCLEO-NFC04A1 shield
- Created a test project using Twister for st25dv

nono313 committed:
drivers: eeprom: add support for st25dv nfc tag

start implementation of eeprom driver for st25dv

Co-Authored-By Martin Schröder <info@swedishembedded.com>

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
boards: shields: add shield x_nucleo_nfc04a1

create new shield x_nucleo_nfc04a1 with eeprom st25dv
and 3 leds

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
tests: drivers: eeprom: write test project for eeprom_st25dv

add tests for specific eeprom_st25dv functions

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
drivers: eeprom: implement specific commands for st25dv

create header for specific st25dv commands
clean up init and add ic_ref command
add enum to check IC ref of st25dv
get end area addresses from st25dv
create specific kconfig file for st25dv
implement additional functions for
st25dv
add function for rf access
protection
read function return 0 if ok
use wait_for macro for timeouts
use assert for check inside driver
use device ready i2c

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

kartben commented:
> * Copied code from original PR

Then please make sure your commits reflect that (`Co-Authored-By`), as well as copyright / license headers in the files which are copied/adapted from original PR.

kartben commented on include/zephyr/drivers/eeprom/eeprom_st25dv.h:
```diff
@@ -0,0 +1,255 @@
+ * @brief Read memory size of the ST25DV
+ *
+ * @param dev the ST25DV device pointer
+ * @param[out] eeprom_st25dv_mem_size_t buffer in which to store the memory size
```
```suggestion
 * @param[out] mem_size buffer in which to store the memory size
```

nono313 replied:
Function documentation updated

nono313 committed:
samples: shields: add sample code for x_nucleo_nfc04a1

sample project exercise some of the st25dv specific features, as well as
standard eeprom API
added error printk in init
update code sample in README of shield sample
use figure to display webp image

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 commented:
Ideally, the test project for the eeprom_st25dv would depends on eeprom driver, but the issue is that this capability is linked to the presence of the shield, and not the board. I have started a discussion on Discord to know how I can specify that the shield is required for the test, without having to add eeprom capability to the nucleo board (which it does not have). If anyone here are an opinion / idea / indication, don't hesitate.

nono313 committed:
boards: shields: update copyright for shield overlay

update copyright
remove usart from overlay
use arduino_gpio requirements for shield
use arduino header pins for LEDs of shield
update device name in shield

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
samples: shields: exclude platforms from sample for shield

exclude platformes not tested that give errors on CI

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
tests: drivers: add fixture for test twister

add fixture to testcase to avoid running on unsupported board

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 commented on tests/drivers/eeprom/eeprom_st25dv/testcase.yaml:
```diff
@@ -0,0 +1,15 @@
+    # Tests overwrite EEPROM content, only run on select boards
+    platform_allow:
+      - nucleo_g0b1re
+    extra_args: SHIELD=x_nucleo_nfc04a1
```
Thanks, I've added a fixture for the normal test (not the build only test)

jfischer-no commented on boards/shields/x_nucleo_nfc04a1/x_nucleo_nfc04a1.overlay:
```diff
@@ -0,0 +1,48 @@
+&arduino_i2c {
+	status="okay";
+
+	eeprom0: eeprom@53 {
```
Please read https://docs.zephyrproject.org/latest/hardware/porting/shields.html#shield-porting-and-configuration

jfischer-no commented on boards/shields/x_nucleo_nfc04a1/Kconfig.defconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Kickmaker
```
The commit message is wrong.

nono313 replied:
Commit message update to 
```boards: shields: add shield x_nucleo_nfc04a1

create new shield x_nucleo_nfc04a1 with eeprom st25dv
and 3 leds

Signed-off-by: Nathan Olff <nathan@kickmaker.net>
```

Is this ok ?

jfischer-no commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -43,7 +43,86 @@ LOG_MODULE_REGISTER(eeprom_st25dv);
+#define ST25DV_REG_I2C_SSO_DYN 0x2004
+
+/** @brief Length of the MEM_SIZE field of the ST25DV */
+#define ST25DV_MEM_SIZE_LENGTH   (2u)
```
There is no reason, unless you do not know what you are doing, to put a single token in parentheses.

nono313 replied:
Old habit sorry. I'm not too familiar (yet) with Zephyr coding rules. Code updated to remove the (uint8_t) on defined and the parentheses

jfischer-no commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -43,7 +43,86 @@ LOG_MODULE_REGISTER(eeprom_st25dv);
+#define EEPROM_ST25DV_WRITE_TIMEOUT 85U
+
+/* I2C_SSO_Dyn */
+#define EEPROM_ST25DV_I2C_SSO_DYN_I2CSSO_SHIFT (uint8_t)(0)
```
Why (uint8_t)?

nono313 replied:
(uint8_t) removed

jfischer-no replied:
Still, why did you want to cast it? if there is a valid reason, this should be written differently. In most cases, 0U or 0UL would be unambiguous.

nono313 replied:
Some of those fields were copied and modified from ST's Linux driver. I tend to use () myself so I left it that way but the cast did not make sense I think.

henrikbrixandersen replied:
> Some of those fields were copied and modified from ST's Linux driver. I tend to use () myself so I left it that way but the cast did not make sense I think.

Please be aware that copying from e.g. GPL licensed code to an Apache-2.0 licensed file is not permitted by the GPL.

nono313 replied:
I have rewritten the copied code and comments. It should be fine now.

jfischer-no reviewed (changes requested):
shield device nodelabel is not following the form <device>_<shield>

kartben commented on samples/shields/x_nucleo_nfc04a1/README.rst:
```diff
@@ -0,0 +1,61 @@
+.. _x-nucleo-nfc04a1-sample:
+
+X-NUCLEO-NFC04A1: shield sample
+#################################################
```
```suggestion
.. zephyr:code-sample:: x-nucleo-nfc04a1
   :name: X-NUCLEO-NFC04A1 shield
   :relevant-api: eeprom_interface gpio_interface

   Interact with the EEPROM and LEDs of an X-NUCLEO-NFC04A1 shield.
```

nono313 replied:
Updated header o the README

kartben commented on samples/shields/x_nucleo_nfc04a1/README.rst:
```diff
@@ -0,0 +1,61 @@
+    X-NUCLEO-NFC04A1 sensor dashboard
+
+
+*** Booting Zephyr OS build 9ad30bf5846e ***
+Found EEPROM device "eeprom@53"
+UUID = F4 01 13 05.
+REV_IC = 12.
+Using eeprom with size of: 4096.
+Device booted 7 times.
+Reset the MCU to see the increasing boot counter
```
incorrect indenting, see https://builds.zephyrproject.io/zephyr/pr/62599/docs/samples/shields/x_nucleo_nfc04a1/README.html

kartben commented on boards/shields/x_nucleo_nfc04a1/doc/index.rst:
```diff
@@ -0,0 +1,69 @@
+.. _x-nucleo-nfc04a1:
+
+X-NUCLEO-NFC04A1: MEMS Inertial and Environmental Multi sensor shield
```
```suggestion
X-NUCLEO-NFC04A1: Dynamic NFC/RFID tag shield
```

nono313 replied:
Updated.

kartben commented on boards/shields/x_nucleo_nfc04a1/doc/img/x-nucleo-nfc04a1.jpg:
please use smaller / cropped webp img -- i made one for you :)  https://gist.github.com/kartben/d9779b8bed55db87220fe4d908e0cc41#file-x-nucleo-nfc04a1-webp

nono313 replied:
Thanks. I didn't know webp was supported. Updated.

nono313 replied:
I seem to get an error on the CI with the webp image. Any idea ? 

```
ERROR: Error in "image" directive:
no content permitted.
```

kartben replied:
s/image/figure/

Figures support captions, images don't. Sorry I should have noticed before asking you for the change :)

kartben commented on boards/shields/x_nucleo_nfc04a1/doc/index.rst:
```diff
@@ -0,0 +1,69 @@
+
+.. image:: img/x-nucleo-nfc04a1.jpg
+     :align: center
+     :alt: X-NUCLEO-NFC04A1
```
```suggestion
     :alt: X-NUCLEO-NFC04A1

     X-NUCLEO-NFC04A1 shield (Credit: STMicroelectronics)
```

nono313 replied:
Thanks, I updated the doc to add the credit

kartben reviewed (changes requested):
mainly doc-related comments, I cannot comment on the actual driver

henrikbrixandersen commented on boards/shields/x_nucleo_nfc04a1/doc/index.rst:
```diff
@@ -0,0 +1,71 @@
+Programming
+***********
+
+The `drivers/eeprom` sample can be used with this shield.
```
Please use a real reference to the sample documentation.

henrikbrixandersen commented on boards/shields/x_nucleo_nfc04a1/Kconfig.defconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Kickmaker
+# SPDX-License-Identifier: Apache-2.0
+
+if SHIELD_X_NUCLEO_NFC04A1
+
+if EEPROM
+
+config EEPROM_ST25DV
+	default y
+
+endif # EEPROM
+
+endif # SHIELD_X_NUCLEO_NFC04A1
```
This should not be needed. `CONFIG_EEPROM_ST25DV` should automatically be enabled based on the shield DTS.

nono313 replied:
I removed the defconfig. It did not work before because I was missing the DT_HAS from https://github.com/zephyrproject-rtos/zephyr/pull/62599#discussion_r1333518581

henrikbrixandersen commented on drivers/eeprom/Kconfig.st25dv:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config EEPROM_ST25DV
+	bool "ST25DV NFC as EEPROM"
+	select I2C
```
```suggestion
	bool "ST25DV NFC as EEPROM"
	default y
	depends on DT_HAS_ST_ST25DV_ENABLED
	select I2C
```

nono313 replied:
Updated

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+#include <zephyr/drivers/eeprom/eeprom_st25dv.h>
+#include <soc.h>
+
+#define LOG_LEVEL CONFIG_EEPROM_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(eeprom_st25dv);
```
```suggestion
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(eeprom_st25dv, CONFIG_EEPROM_LOG_LEVEL);
```

nono313 replied:
Updated. All other eeprom drivers seem to use the same way as I used before.

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+LOG_MODULE_REGISTER(eeprom_st25dv);
+
+/** @brief ST25DV registers */
+typedef enum {
```
Please avoid using typedefs for stuff like this. I'd prefer either just an enum or even better just `#define`.

nono313 replied:
Updated to use #define

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+} st25dv_reg_t;
+
+/** @brief Length of the MEM_SIZE field of the ST25DV */
+#define ST25DV_MEM_SIZE_LENGTH   2u
```
```suggestion
#define ST25DV_MEM_SIZE_LENGTH   2U
```

Please use upper-case `U` all over.

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+ * @details Max value: t_write + t_write * (max write bytes) / (internal page write)
+ * 5 + 5*(256/16)
+ */
+#define EEPROM_ST25DV_WRITE_TIMEOUT 85u
```
```suggestion
#define EEPROM_ST25DV_WRITE_TIMEOUT_MS 85u
```

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+	struct k_mutex lock;
+};
+
+static int _st25dv_read(const struct eeprom_st25dv_config *config, uint16_t addr, off_t offset,
```
Please use the same coding style as in other drivers. E.g.:

```suggestion
static int eeprom_st25dv_read(const struct eeprom_st25dv_config *config, uint16_t addr, off_t offset,
```

nono313 replied:
_st25dv_read is a private function internal to the driver. I already have a eeprom_st25dv_read function that is used for the EEPROM api

```
static const struct eeprom_driver_api eeprom_st25dv_api = {
	.read = eeprom_st25dv_read,
	.write = eeprom_st25dv_write,
	.size = eeprom_st25dv_size,
};
```

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+	 * A write cycle may be in progress so reads must be attempted
+	 * until the current write cycle should be completed.
+	 */
+	timeout = k_uptime_get() + EEPROM_ST25DV_WRITE_TIMEOUT;
+	while (1) {
+		int64_t now = k_uptime_get();
+
+		ret = i2c_write_read(config->bus.bus, addr, wr_data, 2, pdata, count);
+		if (!ret || now > timeout) {
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	}
```
Have you seen the `WAIT_FOR()` macro in `include/zephyr/sys/util.h`?

nono313 replied:
I did not know about the WAIT_FOR macro. I have updated the code to use it. The fact that I want to reuse the return value of the i2c function makes it a bit difficult to read but I guess it's not that harder to read than the manual loop I had earlier

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+
+	k_mutex_init(&data->lock);
+
+	if (!device_is_ready(config->bus.bus)) {
```
Please use `i2c_is_ready_dt()`.

nono313 replied:
Ok I updated the code.

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+ */
+static int _get_enda1(const struct eeprom_st25dv_config *config, uint8_t *const value)
+{
+	if (value == NULL) {
+		return -EINVAL;
+	}
```
Asserts would probably be fine here.

nono313 replied:
Updated to use assert in every NULL check for parameters of function. Is assert() from assert.h the right way to do it ?

henrikbrixandersen commented on drivers/eeprom/eeprom_st25dv.c:
```diff
@@ -0,0 +1,970 @@
+#define ST25DV_DEVICE_INIT(inst)                                                                   \
+	static const struct eeprom_st25dv_config _st25dv_config_##inst = {                         \
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.addr = DT_INST_REG_ADDR(inst),                                                    \
```
What is the use of `.addr`?

nono313 replied:
.addr is used as I2C address when calling i2c driver functions. Sometimes, it is used with | 4 because some command require bit 2 to be set for configuration purposes.

henrikbrixandersen commented on samples/shields/x_nucleo_nfc04a1/src/main.c:
```diff
@@ -0,0 +1,137 @@
+	int rc;
+	int ret;
+
+	if (!gpio_is_ready_dt(&led1) || !gpio_is_ready_dt(&led2) || !gpio_is_ready_dt(&led3)) {
+		return 0;
+	}
+
+	ret = gpio_pin_configure_dt(&led1, GPIO_OUTPUT_ACTIVE);
+	if (ret < 0) {
+		return 0;
+	}
+	ret = gpio_pin_configure_dt(&led2, GPIO_OUTPUT_ACTIVE);
+	if (ret < 0) {
+		return 0;
+	}
+	ret = gpio_pin_configure_dt(&led3, GPIO_OUTPUT_ACTIVE);
+	if (ret < 0) {
+		return 0;
+	}
+
+	if (eeprom == NULL) {
+		return 0;
+	}
```
No error output on the console would make initialization errors unnecessaril hard to debug.

henrikbrixandersen commented on samples/shields/x_nucleo_nfc04a1/src/main.c:
```diff
@@ -0,0 +1,137 @@
+		values.boot_count = 0;
+	}
+
+	values.boot_count++;
+	printk("Device booted %d times.\n", values.boot_count);
```
This is basically replicating the EEPROM boot counter sample. No need for that.

nono313 replied:
Do you mean the sample for the shield should NOT test the EEPROM functionality and only test the ST25DV specific features instead ?

kartben replied:
True -- but the shield samples seem to be (not saying it's necessarily the right approach) more about showcasing the capabilities of said shields, and this is particularly true for Nucleo shields (and yes, that also means they probably duplicate functionality from existing samples). I guess these are more "demos" than samples.

henrikbrixandersen reviewed (changes requested):
While I can see why you went ahead and implemented this as an "extended" EEPROM driver, I think there is way too much non-EEPROM related code and functionality in this EEPROM driver. The fact that you felt the need for a driver-specific test instead of the standard EEPROM API test supports this.

I would suggest looking into Multi-Function Device drivers (`drivers/mfd`) for this device. The MFD driver could host the basic device access and a dedicated EEPROM and a dedicated NFC could be built on top of that.

I would very much welcome an experimental NFC API in Zephyr, but tucking a device-specific NFC API in under an EEPROM driver is not acceptable, sorry.

Please also see https://github.com/zephyrproject-rtos/zephyr/issues/5703

kartben commented on samples/shields/x_nucleo_nfc04a1/README.rst:
```diff
@@ -0,0 +1,61 @@
+Sample Output
+=============
+
+ .. code-block:: console
```
Bad formatting of the code block, see https://builds.zephyrproject.io/zephyr/pr/62599/docs/samples/shields/x_nucleo_nfc04a1/README.html

nono313 commented:
@henrikbrixandersen Sorry I did not knew about MFD. I knew when I started that ther was no NFC API in upstream and that the issue of building it was still open since 2018. The latest comments on the NFC issue happened only 2 days ago, which is after I started working on this driver.

The NFC API proposal seem very complete and contain a lot of things I wasn't planning on implementing (like NDEF, data exchange...). Features like memory area configuration and RF access protection configuration on the other side does not seem to be included. I'm not very familiar with NFC and don't really know if these features I need are standard things for NFC / EEPROM chips or specific to the ST25DV so I don't know if things like these should be added to NFC API (I guess not).

Do you suggest I should close this PR and just wait for the experimental NFC api to be included ? As I said there seem to be a lot in this driver proposal that I don't have an am not sure how to implement anyway...

henrikbrixandersen commented:
> @henrikbrixandersen Sorry I did not knew about MFD. I knew when I started that ther was no NFC API in upstream and that the issue of building it was still open since 2018. The latest comments on the NFC issue happened only 2 days ago, which is after I started working on this driver.
> 
> The NFC API proposal seem very complete and contain a lot of things I wasn't planning on implementing (like NDEF, data exchange...). Features like memory area configuration and RF access protection configuration on the other side does not seem to be included. I'm not very familiar with NFC and don't really know if these features I need are standard things for NFC / EEPROM chips or specific to the ST25DV so I don't know if things like these should be added to NFC API (I guess not).
> 
> Do you suggest I should close this PR and just wait for the experimental NFC api to be included ? As I said there seem to be a lot in this driver proposal that I don't have an am not sure how to implement anyway...

Another option could be to work with the authors of the proposed NFC API to ensure it will meet the requirements for implementing an ST25DV NFC driver. You could convert this PR to draft mode or close it while working on this.

RomainPelletant commented:
@henrikbrixandersen @nono313 IMHO, NFC communication represents 10% of features that EEPROM offers.
On the other hand, that EEPROM offers the ability to set permissions on data part for NFC communication and a great feature (purely EEPROM-oriented) could be partitioning with NFC flags permission:

`&st25dv {
    partitions {
        compatible = "fixed-partitions";
        #address-cells = <1>;
        #size-cells = <1>;

	interface_loader_partition: partition@0 {
		label = "interface_loader";
		reg = <0x00000000 0x00000040>;
		nfc_read_only;
	};
	device_data_partition: partition@40 {
		label = "device_data";
		reg = <0x00000040 DT_SIZE_K(1)>;
		nfc_read_write;
	};
};
`

Do you think it is an interesting feature to add?

henrikbrixandersen commented:
> Do you think it is an interesting feature to add?

Not sure exactly what you are suggesting here, but my earlier comment still stands: a separate API is needed for NFC support. Driver-specific API extensions to an EEPROM driver for adding NFC is too much of a hack for me to accept it.

nono313 committed:
drivers: eeprom: implement access to mailbox

implement access to dynamic register and mailbox feature for fast
transfer mode

add limited i2c attribute for when host I2C is not able to write 256
bytes

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
tests: drivers: add st25dv to build all tests

add st25dv driver to build_all eeprom tests

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
drivers: eeprom: replace wait_for macros

wait_for macros do not work as they execute the expression twice

Signed-off-by: Nathan Olff <nathan@kickmaker.net>

nono313 committed:
drivers: eeprom: use compiler guard for advanced features of st25dv

add kconfig for enabling / disabling advanced features of the st25dv
(other than basic eeprom functionality)

Signed-off-by: Nathan Olff <nathan@kickmaker.net>
